source/AMDSThreadedTCPDataServer.cpp

```diff
@@ -0,0 +1,50 @@
+#include "source/AMDSThreadedTCPDataServer.h"
+
+#include <QCoreApplication>
+
+AMDSThreadedTCPDataServer::AMDSThreadedTCPDataServer(QString hwType, QObject *parent) :
+	QObject(parent)
+{
+	hwType_ = hwType;
+
+	thread_ = new QThread;
+	server_ = new AMDSTCPDataServer;
+
+	connect(thread_, SIGNAL(started()), this, SLOT(onThreadStarted()));
+	connect(this, SIGNAL(startServer(QString,quint16)), server_, SLOT(start(QString,quint16)));
+	connect(server_, SIGNAL(error(quint8,quint16,QString)), this, SIGNAL(error(quint8,quint16,QString)));
+
+	server_->moveToThread(thread_);
+	thread_->start(QThread::LowPriority);
+}
+
+AMDSThreadedTCPDataServer::~AMDSThreadedTCPDataServer()
+{
+	if (thread_->isRunning())
+		thread_->terminate();
```

Comment:
This needs to be changed to thread_->quit().
quit() is the graceful close, terminate() is the 'oh crap!' emergency exit. From the QThread::terminate() documentation:

Warning: This function is dangerous and its use is discouraged. The thread can be terminated at any point in its code path. Threads can be terminated while modifying data. There is no chance for the thread to clean up after itself, unlock any held mutexes, etc. In short, use this function only if absolutely necessary.